cmds/netcat/netcat.go

```diff
@@ -1,37 +1,182 @@
+	case status := <-c2:
+		log.Printf("Local program is terminated: %+v\n", status)
 	}
+}
 
-	if c, err = net.Dial("tcp", os.Args[1]); err != nil {
-		fmt.Printf("%v\n", err)
-		os.Exit(1)
+// Launch receive goroutine first, wait for address from it (if needed),
+// launch send goroutine then.
+func transferPackets(con net.Conn) {
+	c1 := readAndWrite(con, os.Stdout)
+	// If connection hasn't got remote address then wait for it from receiver goroutine
+	ra := con.RemoteAddr()
+	if ra == nil {
+		progress := <-c1
+		ra = progress.ra
+		log.Println("Connect from", ra)
+	}
+	c2 := readAndWriteToAddr(os.Stdin, con, ra)
+	select {
+	case progress := <-c1:
+		log.Printf("Remote connection is closed: %+v\n", progress)
+	case progress := <-c2:
+		log.Printf("Local program is terminated: %+v\n", progress)
 	}
+}
+
+// Read from Reader and write to Writer until EOF.
+func readAndWrite(r io.Reader, w io.Writer) <-chan progress {
+	return readAndWriteToAddr(r, w, nil)
+}
+
+// Read from Reader and write to Writer until EOF.
+// ra is an address to whom packets must be sent in UDP listen mode.
+func readAndWriteToAddr(r io.Reader, w io.Writer, ra net.Addr) <-chan progress {
+	buf := make([]byte, bufferLimit)
+	c := make(chan progress)
+	rBytes, wBytes := uint64(0), uint64(0)
 	go func() {
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			fmt.Printf("%v", err)
+		defer func() {
+			if con, ok := w.(net.Conn); ok {
+				con.Close()
+				if _, ok := con.(*net.UDPConn); ok {
+					log.Printf("Stop receiving flow from %v\n", ra)
+				} else {
+					log.Printf("Connection from %v is closed\n", con.RemoteAddr())
+				}
+			}
+			c <- progress{rBytes: rBytes, wBytes: wBytes, ra: ra}
+		}()
+
+		for {
+			var n int
+			var err error
+
+			// Read
+			if con, ok := r.(*net.UDPConn); ok {
+				var addr net.Addr
+				n, addr, err = con.ReadFrom(buf)
+				// Inform caller function with remote address once
+				// (for UDP in listen mode only)
+				if con.RemoteAddr() == nil && ra == nil {
+					ra = addr
+					c <- progress{rBytes: rBytes, wBytes: wBytes, ra: ra}
+				}
+			} else {
+				n, err = r.Read(buf)
+			}
+			if err != nil {
+				if err != io.EOF {
+					log.Printf("Read error: %s\n", err)
+				}
+				break
+			}
+			rBytes += uint64(n)
+
+			// Write
+			if con, ok := w.(*net.UDPConn); ok && con.RemoteAddr() == nil {
+				// Special case for UDP in listen mode otherwise
+				// net.ErrWriteToConnected will be thrown
+				n, err = con.WriteTo(buf[0:n], ra)
+			} else {
+				n, err = w.Write(buf[0:n])
+			}
+			if err != nil {
+				log.Fatalf("Write error: %s\n", err)
+			}
+			wBytes += uint64(n)
 		}
 	}()
-	if _, err = io.Copy(os.Stdout, c); err != nil {
-		fmt.Printf("%v", err)
+	return c
+}
+
+func main() {
+	var (
+		host   = flag.String("host", "", "Remote host to connect, i.e. 127.0.0.1")
+		proto  = flag.String("proto", "tcp", "TCP/UDP mode")
+		port   = flag.String("port", ":9999", "Port to listen on or connect to (prepended by colon), i.e. :9999")
```

Comment:
is this format for a port standard in POSIX or elsewhere (prepending with a colon)? if we're not standard compliant we can eliminate the port altogether and use -host="localhost:port". this is a minor point.